hebiros_gazebo_plugin/sim/joint.cpp

```diff
@@ -0,0 +1,347 @@
+static constexpr double DEFAULT_VELOCITY_KP = 0.05;
+static constexpr double DEFAULT_VELOCITY_KI = 0.0;
+static constexpr double DEFAULT_VELOCITY_KD = 0.0;
+static constexpr double DEFAULT_VELOCITY_FF = 1.0;
+static constexpr double DEFAULT_EFFORT_KP = 0.25;
+static constexpr double DEFAULT_EFFORT_KI = 0.0;
+static constexpr double DEFAULT_EFFORT_KD = 0.001;
+static constexpr double DEFAULT_EFFORT_FF = 1.0;
+
+static constexpr double GEAR_RATIO_X5_1 = 272.22;
+static constexpr double GEAR_RATIO_X8_3 = 272.22;
+static constexpr double GEAR_RATIO_X5_4 = 762.22;
+static constexpr double GEAR_RATIO_X8_9 = 762.22;
+static constexpr double GEAR_RATIO_X5_9 = 1742.22;
+static constexpr double GEAR_RATIO_X8_16 = 1462.222;
+
+static std::map<Joint::JointType, double> gear_ratios = {
+  {Joint::JointType::X5_1, 272.22},
+  {Joint::JointType::X5_4, 762.22},
+  {Joint::JointType::X5_9, 1742.22},
+  {Joint::JointType::X8_3, 272.22},
+  {Joint::JointType::X8_9, 762.22},
+  {Joint::JointType::X8_16, 1462.222}
+};
+
+double getVelocityFF(double gear_ratio, bool is_x8) {
+  double voltage = 48;
+  double speed_constant = is_x8 ? 1360 : 1530;
+  return gear_ratio / (speed_constant * (2.0 * M_PI / 60.0) * voltage);
+}
+
+double getEffortFF(double gear_ratio, bool is_x8) {
+  double voltage = 48;
+  double term_resistance = is_x8 ? 1360 : 1530;
+  double torque_constant = is_x8 ? (9.32f / 1000.0f) : (6.26f / 1000.0f);
+  return term_resistance / (gear_ratio * voltage * torque_constant * 0.65);
+}
+
+/////////////////////////////////
+// END REFACTOR TODO
+/////////////////////////////////
+  
+std::unique_ptr<Joint> Joint::tryCreate(const std::string& family, const std::string& name, const std::string& type)
+{
+  // Set joint type:
+  JointType joint_type;
+  if (type == "X5_1")
+    joint_type = JointType::X5_1;
+  else if (type == "X5_4")
+    joint_type = JointType::X5_4;
+  else if (type == "X5_9")
+    joint_type = JointType::X5_9;
+  else if (type == "X8_3")
+    joint_type = JointType::X8_3;
+  else if (type == "X8_9")
+    joint_type = JointType::X8_9;
+  else if (type == "X8_16")
+    joint_type = JointType::X8_16;
+  else
+    return nullptr;
+
+  // Create module if it is a valid type
+  return std::unique_ptr<Joint>(new Joint(family + "/" + name, joint_type));
+}
+
+bool isX8(Joint::JointType jt)
+{
+  using JT = Joint::JointType;
+  return jt == JT::X8_3 || jt == JT::X8_9 || jt == JT::X8_16;
+}
+
+Joint::Joint(const std::string& name,
+  JointType joint_type)
+  : joint_type_(joint_type),
+    gear_ratio_(gear_ratios.at(joint_type)),
+    temperature_(isX8(joint_type) ?
+      hebi::sim::TemperatureModel::createX8() :
+      hebi::sim::TemperatureModel::createX5()),
+    name_(name),
+    position_pid_(1),
+    velocity_pid_(getVelocityFF(gear_ratio_, isX8(joint_type_))),
+    effort_pid_(getEffortFF(gear_ratio_, isX8(joint_type_))) {
+
+  // Set default gains:
+  // TODO: cleaner way to do this? Map structure?
```

Comment:
Lol, you're not kidding...

I'm leaning towards an xml file here. Maybe a map could suffice too, but why not just let it sit in a file, so anyone can hack around in it for whatever reason.

Reply:
Ideally, we would just store the default gains files downloaded from our website, and load those in.  I don't want to get into the xml parser code yet here, though...